File: tools/census-harvest/tests/fixtures/ts-mini/src/engine.ts
import { Stats, Store } from "./store";

export async function compact(store: Store, key: string): Promise<number> {
  const data = await store.getObject(key);
  if (data === null) {
    throw new Error("missing object");
  }
  store.putObject(key, data);
  return store.stats().objects;
}

export function planCompaction(store: Store): string[] {
  return [store.cfg.region];
}

function estimate(store: Store): number {
  return store.cfg.maxSize;
}

export function mergeRanges(
  store: Store,
  lo: number,
  hi: number,
): number {
  if (lo > hi) {
    throw new Error("bad range");
  }
  return estimate(store) + hi - lo;
}

export function report(store: Store): Stats {
  return store.stats();
}

File: tools/census-harvest/tests/fixtures/rust-mini/src/engine.rs
use crate::store::{Persist, Stats, Store};

pub fn compact(store: &mut Store, key: &str) -> Result<u64, String> {
    let data = store.get_object(key)?;
    store.put_object(key, data)?;
    store.persist(key)?;
    Ok(store.stats().objects)
}

pub(crate) fn plan_compaction(store: &Store) -> Vec<String> {
    vec![store.cfg.region.clone()]
}

fn estimate(store: &Store) -> usize {
    store.cfg.max_size
}

pub fn merge_ranges(
    store: &Store,
    lo: usize,
    hi: usize,
) -> Result<usize, String> {
    if lo > hi {
        return Err("bad range".to_string());
    }
    Ok(estimate(store) + hi - lo)
}

pub fn report(store: &Store) -> Stats {
    store.stats()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::store::Config;

    fn compact_helper(store: &mut Store) -> u64 {
        store.stats().objects
    }

    #[test]
    fn test_compact() {
        let mut s = Store::new(Config { max_size: 8, region: "eu".to_string() });
        let n = compact(&mut s, "k").unwrap();
        assert_eq!(n, compact_helper(&mut s));
    }
}
